### PR TITLE
Close #13766 : Refactor Clear Scenery to class

### LINKED ISSUE #13766: Refactor window to class: ClearScenery
See #13737 for more details.

### PR DESCRIPTION
**Pull request by Bepso**

### CODE REVIEW

File: src/openrct2-ui/interface/Window.cpp [hunk 1]
```diff
@@ -751,3 +752,14 @@ void Window::Close()
 {
     window_close(this);
 }
+
+void Window::WindowTextInputOpen(rct_widgetindex call_widget,
```
Comment (duncanspumpkin) [inline src/openrct2-ui/interface/Window.cpp:756]: const as many things as possible.
Coding style is camelCase for variables.

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 2]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+        {
+            Formatter ft;
+            
+            switch (widgetIndex)
+                {
+                case WIDX_CLOSE:
+                    Close();
+                    break;
+                case WIDX_PREVIEW:
+                    ft.Add<int16_t>(MINIMUM_TOOL_SIZE);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:85]: Declare ft only in this case.

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 3]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+                        // Invalidate the window
+                        Invalidate();
+                        break;
+                }
+        }
+
+        void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
+        {
+            int32_t size;
+            char* end;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:128]: do not forward declare this isn't C89 code anymore.

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 4]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
 
-static void window_clear_scenery_inputsize(rct_window* w)
+rct_window* window_clear_scenery_open()
 {
-    Formatter ft;
-    ft.Add<int16_t>(MINIMUM_TOOL_SIZE);
-    ft.Add<int16_t>(MAXIMUM_TOOL_SIZE);
-    window_text_input_open(w, WIDX_PREVIEW, STR_SELECTION_SIZE, STR_ENTER_SELECTION_SIZE, ft, STR_NONE, STR_NONE, 3);
-}
+    auto w = static_cast<CleanSceneryWindow*>(window_bring_to_front_by_class(WC_CLEAR_SCENERY));
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:190]: auto*

**Comment by duncanspumpkin** [comment]
You'll need to clang format your code.

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 5]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+            gClearFootpath = false;
+        }
+
+        void OnClose() override
+        {
+            if (clear_scenery_tool_is_active())
+                tool_cancel();
+        }
+
+        void OnMouseUp(rct_widgetindex widgetIndex) override
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:75]: Go through and add const on pretty much every parameter in every function call of this window. (Everything except OnDraw which does modify its parameter)

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 6]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+                        break;
+                }
+        }
+
+        void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
+        {
+            int32_t size;
+            char* end;
+
+            if (widgetIndex != WIDX_PREVIEW || text == nullptr)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:130]: std::string_view cant be compared to `nullptr` think it should be `text.empty()`

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 7]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+                return;
+
+            size = strtol(text.data(), &end, 10);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:133]: This is not safe to perform you will need to convert the string view into a string. By doing this you are ignoring the length of the view.

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 8]
```diff
@@ -33,208 +33,169 @@ static constexpr const rct_string_id WINDOW_TITLE = STR_CLEAR_SCENERY;
...
+            if (*end == '\0')
+            {
+                size = std::max(MINIMUM_TOOL_SIZE, size);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/ClearScenery.cpp:136]: `std::clamp(size, MINIMUM_TOOL_SIZE, MAXIMUM_TOOL_SIZE);`

File: src/openrct2-ui/interface/Window.h [hunk 9]
```diff
@@ -26,6 +26,13 @@ struct Window : rct_window
     void SetCheckboxValue(rct_widgetindex widgetIndex, bool value);
     void DrawWidgets(rct_drawpixelinfo& dpi);
     void Close();
+    void WindowTextInputOpen(rct_widgetindex call_widget,
```
Comment (duncanspumpkin) [inline src/openrct2-ui/interface/Window.h:29]: Why are you adding this as a function of the Window?

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
See comments.

File: src/openrct2-ui/windows/ClearScenery.cpp (hunk 6 above)
Comment (Bepso) [reply src/openrct2-ui/windows/ClearScenery.cpp]: This was copied from the original code, which had the `text == nullptr.` But i changed it to `text.empty()` now anyways

File: src/openrct2-ui/interface/Window.h [hunk 10]
```diff
@@ -26,6 +26,13 @@ struct Window : rct_window
     void SetCheckboxValue(rct_widgetindex widgetIndex, bool value);
     void DrawWidgets(rct_drawpixelinfo& dpi);
     void Close();
+    void TextInputOpen(rct_widgetindex call_widget,
+                             rct_string_id title,
+                             rct_string_id description,
+                             const Formatter& descriptionArgs,
+                             rct_string_id existing_text,
```
Comment (duncanspumpkin) [inline src/openrct2-ui/interface/Window.h:33]: still have snake case for these two parameters.

**Review by duncanspumpkin — APPROVED** [review]
Apart from those two parameters this is good to go.

File: src/openrct2-ui/interface/Window.h (hunk 10 above)
Comment (Bepso) [reply src/openrct2-ui/interface/Window.h]: Haha, that is my bad. Updated that right now

**Comment by Bepso** [comment]
Hopefully I have fixed all the clang format errors and the commit messages now

**Comment by duncanspumpkin** [comment]
In future never merge in develop always rebase. It's fine in this instance as we will squash merge but keep it in mind for future

**Comment by Bepso** [comment]
Yes, that is my fault. Thank you!

**Pull request merged**